# Wrangle Pisa 2012
#### By Gabriela Sikora


### Abstract



### Table of Contents

<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
    <ul>
    <li><a href="#gathering">Gathering Data</a></li>
    <li><a href="#assessing1">Initial Assessment</a></li>
        <ul>
          <li><a href="#issues1">Identified Issues</a></li>
        </ul>
    <li><a href="#cleaning1">Cleaning Data</a></li>
    <li><a href="#assessing2">Final Assessment</a></li>
    </ul>
<li><a href="#sav">Store</a></li>
<li><a href="#conclusion">Conclusion</a></li>
<li><a href="#references">References</a></li>    
</ul>

<a id='intro'></a>
## Introduction



To begin, we first need to import all the relevant libraries:
#### Import

In [1]:
# import all packages
import numpy as np
import pandas as pd

<a id='wrangling'></a>
## Data Wrangling

In this section of the report, we will load in the data, check it for cleanliness, and then trim and clean the datasets for analysis. 

<a id='gathering'></a>
### Gathering Data

Unfortunately the Pisa dataset is too large to be uploaded to github, but if you would like to work with it yourself then place follow this link: https://s3.amazonaws.com/udacity-hosted-downloads/ud507/pisa2012.csv.zip 

In [2]:
# Encoding help found here: https://stackoverflow.com/questions/30462807/encoding-error-in-panda-read-csv

# Load in the Pisa dataframe
pisa = pd.read_csv('pisa2012.csv', encoding="cp1252", sep=',')

# Load in the Pisa dictionary
pisa_dict = pd.read_csv('pisadict2012.csv', encoding="cp1252", sep=',')

/Users/gabriela/Desktop/Code/myenv/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3044: DtypeWarning: Columns (15,16,17,21,22,23,24,25,26,30,31,36,37,45,65,123,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,284,285,286,287,288,289,290,291,292,293,294,295,296,297,298,299,300,301,302,303,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,376,377,378,379,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,398,399,400,401,402,403,475) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


<a id='assessing1'></a>
### First Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [3]:
pisa

Unnamed: 0      CNT  SUBNATIO  STRATUM      OECD         NC  SCHOOLID  \
0                1  Albania     80000  ALB0006  Non-OECD    Albania         1   
1                2  Albania     80000  ALB0006  Non-OECD    Albania         1   
2                3  Albania     80000  ALB0006  Non-OECD    Albania         1   
3                4  Albania     80000  ALB0006  Non-OECD    Albania         1   
4                5  Albania     80000  ALB0006  Non-OECD    Albania         1   
5                6  Albania     80000  ALB0006  Non-OECD    Albania         1   
6                7  Albania     80000  ALB0006  Non-OECD    Albania         1   
7                8  Albania     80000  ALB0006  Non-OECD    Albania         1   
8                9  Albania     80000  ALB0006  Non-OECD    Albania         1   
9               10  Albania     80000  ALB0005  Non-OECD    Albania         2   
10              11  Albania     80000  ALB0005  Non-OECD    Albania         2   
11              12  Albania     80000  ALB0005  Non-OECD    Albania         2   
12              13  Albania     80000  ALB0005  Non-OECD    Albania         2   
13              14  Albania     80000  ALB0005  Non-OECD    Albania         2   
14              15  Albania     80000  ALB0005  Non-OECD    Albania         2   
15              16  Albania     80000  ALB0005  Non-OECD    Albania         2   
16              17  Albania     80000  ALB0005  Non-OECD    Albania         2   
17              18  Albania     80000  ALB0005  Non-OECD    Albania         2   
18              19  Albania     80000  ALB0005  Non-OECD    Albania         2   
19              20  Albania     80000  ALB0005  Non-OECD    Albania         2   
20              21  Albania     80000  ALB0005  Non-OECD    Albania         2   
21              22  Albania     80000  ALB0005  Non-OECD    Albania         2   
22              23  Albania     80000  ALB0005  Non-OECD    Albania         2   
23              24  Albania     80000  ALB0005  Non-OECD    Albania         2   
24              25  Albania     80000  ALB0005  Non-OECD    Albania         2   
25              26  Albania     80000  ALB0005  Non-OECD    Albania         2   
26              27  Albania     80000  ALB0005  Non-OECD    Albania         2   
27              28  Albania     80000  ALB0005  Non-OECD    Albania         2   
28              29  Albania     80000  ALB0005  Non-OECD    Albania         2   
29              30  Albania     80000  ALB0005  Non-OECD    Albania         2   
...            ...      ...       ...      ...       ...        ...       ...   
485460      485461  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485461      485462  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485462      485463  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485463      485464  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485464      485465  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485465      485466  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485466      485467  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485467      485468  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485468      485469  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485469      485470  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485470      485471  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485471      485472  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485472      485473  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485473      485474  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485474      485475  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485475      485476  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485476      485477  Vietnam   7040000  VNM0317  Non-OECD  Viet Nam        162   
485477      485478  Vietnam   7040000  

In [4]:
pisa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Columns: 636 entries, Unnamed: 0 to VER_STU
dtypes: float64(250), int64(18), object(368)
memory usage: 2.3+ GB


In [5]:
pisa_dict

Unnamed: 0                                                  x
0           CNT                           Country code 3-character
1      SUBNATIO  Adjudicated sub-region code 7-digit code (3-di...
2       STRATUM  Stratum ID 7-character (cnt + region ID + orig...
3          OECD                                       OECD country
4            NC                       National Centre 6-digit Code
5      SCHOOLID  School ID 7-digit (region ID + stratum ID + 3-...
6       STIDSTD                                         Student ID
7       ST01Q01                                International Grade
8       ST02Q01                           National Study Programme
9       ST03Q01                                      Birth - Month
10      ST03Q02                                        Birth -Year
11      ST04Q01                                             Gender
12      ST05Q01                                   Attend <ISCED 0>
13      ST06Q01                                   Age at <ISCED 1>
14      ST07Q01                                 Repeat - <ISCED 1>
15      ST07Q02                                 Repeat - <ISCED 2>
16      ST07Q03                                 Repeat - <ISCED 3>
17      ST08Q01                          Truancy - Late for School
18      ST09Q01                    Truancy - Skip whole school day
19     ST115Q01           Truancy - Skip classes within school day
20      ST11Q01                                   At Home - Mother
21      ST11Q02                                   At Home - Father
22      ST11Q03                                 At Home - Brothers
23      ST11Q04                                  At Home - Sisters
24      ST11Q05                             At Home - Grandparents
25      ST11Q06                                   At Home - Others
26      ST13Q01                          Mother<Highest Schooling>
27      ST14Q01            Mother Qualifications - <ISCED level 6>
28      ST14Q02           Mother Qualifications - <ISCED level 5A>
29      ST14Q03           Mother Qualifications - <ISCED level 5B>
..          ...                                                ...
605    W_FSTR55           FINAL STUDENT REPLICATE BRR-FAY WEIGHT55
606    W_FSTR56           FINAL STUDENT REPLICATE BRR-FAY WEIGHT56
607    W_FSTR57           FINAL STUDENT REPLICATE BRR-FAY WEIGHT57
608    W_FSTR58           FINAL STUDENT REPLICATE BRR-FAY WEIGHT58
609    W_FSTR59           FINAL STUDENT REPLICATE BRR-FAY WEIGHT59
610    W_FSTR60           FINAL STUDENT REPLICATE BRR-FAY WEIGHT60
611    W_FSTR61           FINAL STUDENT REPLICATE BRR-FAY WEIGHT61
612    W_FSTR62           FINAL STUDENT REPLICATE BRR-FAY WEIGHT62
613    W_FSTR63           FINAL STUDENT REPLICATE BRR-FAY WEIGHT63
614    W_FSTR64           FINAL STUDENT REPLICATE BRR-FAY WEIGHT64
615    W_FSTR65           FINAL STUDENT REPLICATE BRR-FAY WEIGHT65
616    W_FSTR66           FINAL STUDENT REPLICATE BRR-FAY WEIGHT66
617    W_FSTR67           FINAL STUDENT REPLICATE BRR-FAY WEIGHT67
618    W_FSTR68           FINAL STUDENT REPLICATE BRR-FAY WEIGHT68
619    W_FSTR69           FINAL STUDENT REPLICATE BRR-FAY WEIGHT69
620    W_FSTR70           FINAL STUDENT REPLICATE BRR-FAY WEIGHT70
621    W_FSTR71           FINAL STUDENT REPLICATE BRR-FAY WEIGHT71
622    W_FSTR72           FINAL STUDENT REPLICATE BRR-FAY WEIGHT72
623    W_FSTR73           FINAL STUDENT REPLICATE BRR-FAY WEIGHT73
624    W_FSTR74           FINAL STUDENT REPLICATE BRR-FAY WEIGHT74
625    W_FSTR75           FINAL STUDENT REPLICATE BRR-FAY WEIGHT75
626    W_FSTR76           FINAL STUDENT REPLICATE BRR-FAY WEIGHT76
627    W_FSTR77           FINAL STUDENT REPLICATE BRR-FAY WEIGHT77
628    W_FSTR78           FINAL STUDENT REPLICATE BRR-FAY WEIGHT78
629    W_FSTR79           FINAL STUDENT REPLICATE BRR-FAY WEIGHT79
630    W_FSTR80           FINAL STUDENT REPLICATE BRR-FAY WEIGHT80
631    WVARSTRR           RANDOMIZED FINAL VARIANCE STRATUM (1-80)
632    VAR_UNIT                    RANDOMLY ASSIGNED

In [6]:
# Check for null values in relevant columns
columns = ['STIDSTD', 'ST01Q01', 'ST04Q01', 'ST13Q01', 
            'ST17Q01', 'ST28Q01', 'ST57Q01', 'ST57Q02', 'ST57Q03', 
            'ST57Q04', 'ST57Q05', 'MMINS', 'LMINS', 'SMINS',
            'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',  
            'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
            'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']

for x in columns:
    print(x, pisa[x].isna().sum())

STIDSTD 0
ST01Q01 0
ST04Q01 0
ST13Q01 27511
ST17Q01 42229
ST28Q01 11725
ST57Q01 184123
ST57Q02 215682
ST57Q03 201677
ST57Q04 205833
ST57Q05 195988
MMINS 202187
LMINS 202624
SMINS 214576
PV1MATH 0
PV2MATH 0
PV3MATH 0
PV4MATH 0
PV5MATH 0
PV1READ 0
PV2READ 0
PV3READ 0
PV4READ 0
PV5READ 0
PV1SCIE 0
PV2SCIE 0
PV3SCIE 0
PV4SCIE 0
PV5SCIE 0


In [7]:
pisa['CNT'].value_counts()

Mexico                      33806
Italy                       31073
Spain                       25313
Canada                      21544
Brazil                      19204
Australia                   14481
United Kingdom              12659
United Arab Emirates        11500
Switzerland                 11229
Qatar                       10966
Colombia                     9073
Finland                      8829
Belgium                      8597
Denmark                      7481
Jordan                       7038
Chile                        6856
Thailand                     6606
Japan                        6351
Chinese Taipei               6046
Peru                         6035
Slovenia                     5911
Argentina                    5908
Kazakhstan                   5808
Portugal                     5722
Indonesia                    5622
Singapore                    5546
Macao-China                  5335
Czech Republic               5327
Uruguay                      5315
Bulgaria      

In [8]:
pisa.duplicated().sum()

0

<a id='issues1'></a>
### Identified Issues
Here we can see some of the issues noticed in the dataframes. These quality issues will be cleaned in the following section.

1. Narrow down the dataframe by only looking at variables of interest since there are more variables than needed
2. Make the variable names readable since they are codes that can only be read with the dictionary
3. Create an average of all Math, Reading, and Science scores since all 5 scores for each are not necessary
4. Missing values in relevant columns ('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')
5. Create a total for Out-of-School Study Time and Learning time (minutes per week)
6. Single countries have various entries in the country column (Florida (USA), Massachusetts (USA), Connecticut (USA) and United States of America)


<a id='cleaning1'></a>
### First Cleaning

In [9]:
# Create a copy to preserve the original dataframe
pisa_clean = pisa.copy()

### Narrow down the dataframe by only looking at variables of interest

There are 636 columns available in the original dataframe, which is far more than what is necessary. We will rather look at just the following variables.

In [10]:
pisa_clean = pisa_clean[['STIDSTD', 'ST04Q01', 'ST13Q01', 'ST14Q01', 'ST14Q02', 'ST14Q03', 'ST14Q04', 
                         'ST17Q01', 'ST18Q01', 'ST18Q02', 'ST18Q03', 'ST18Q04', 
                         'ST28Q01', 
                         'ST57Q01', 'ST57Q02', 'ST57Q03', 'ST57Q04', 'ST57Q05', 'MMINS', 
                         'LMINS', 'SMINS', 'PV1MATH', 'PV2MATH', 'PV3MATH', 'PV4MATH', 
                         'PV5MATH', 'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ', 'PV5READ',
                         'PV1SCIE', 'PV2SCIE', 'PV3SCIE', 'PV4SCIE', 'PV5SCIE']]       

In [11]:
# Check to see if the columns are correctly reduced
list(pisa_clean)

['STIDSTD',
 'ST04Q01',
 'ST13Q01',
 'ST14Q01',
 'ST14Q02',
 'ST14Q03',
 'ST14Q04',
 'ST17Q01',
 'ST18Q01',
 'ST18Q02',
 'ST18Q03',
 'ST18Q04',
 'ST28Q01',
 'ST57Q01',
 'ST57Q02',
 'ST57Q03',
 'ST57Q04',
 'ST57Q05',
 'MMINS',
 'LMINS',
 'SMINS',
 'PV1MATH',
 'PV2MATH',
 'PV3MATH',
 'PV4MATH',
 'PV5MATH',
 'PV1READ',
 'PV2READ',
 'PV3READ',
 'PV4READ',
 'PV5READ',
 'PV1SCIE',
 'PV2SCIE',
 'PV3SCIE',
 'PV4SCIE',
 'PV5SCIE']

### Create an average of all Math, Reading, and Science scores

For each of the Math, Reading, and Science scores, there are 5 scores provided. The variation between these scores is not needed, so rather we will look at the average for each of these major categories and remove the individual scores.

In [12]:
# Create average of Math scores
pisa_clean['Average Math Score']    = (pisa_clean['PV1MATH'] + pisa_clean['PV2MATH'] + pisa_clean['PV3MATH'] + pisa_clean['PV4MATH'] + pisa_clean['PV5MATH']) / 5

# Create average of Reading scores
pisa_clean['Average Reading Score'] = (pisa_clean['PV1READ'] + pisa_clean['PV2READ'] + pisa_clean['PV3READ'] + pisa_clean['PV4READ'] + pisa_clean['PV5READ']) / 5

# Create average of Science scores
pisa_clean['Average Science Score'] = (pisa_clean['PV1SCIE'] + pisa_clean['PV2SCIE'] + pisa_clean['PV3SCIE'] + pisa_clean['PV4SCIE'] + pisa_clean['PV5SCIE']) / 5

# Create average score
pisa_clean['Average Total Score']   = (pisa_clean['Average Math Score'] + pisa_clean['Average Reading Score'] + pisa_clean['Average Science Score'] +pisa_clean['Average Math Score']) / 3

In [13]:
# Remove the individual scores for Math, Reading, and Science
pisa_clean.drop(columns=['PV1MATH',  'PV2MATH', 'PV3MATH', 'PV4MATH', 'PV5MATH',   
                         'PV1READ', 'PV2READ',  'PV3READ', 'PV4READ',  'PV5READ',
                         'PV1SCIE',  'PV2SCIE', 'PV3SCIE',  'PV4SCIE', 'PV5SCIE'], inplace=True)

In [14]:
# Test to see accuracy of the Math score
print(pisa_clean['Average Math Score'].describe())
print('-'*22)

# Test to see accuracy of the Reading score
print(pisa_clean['Average Reading Score'].describe())
print('-'*22)

# Test to see accuracy of the Science score
print(pisa_clean['Average Science Score'].describe())

# Test to see accuracy of the Total score
print(pisa_clean['Average Total Score'].describe())

count    485490.000000
mean        469.651234
std         100.786610
min          54.767080
25%         396.019620
50%         465.734520
75%         540.123060
max         903.107960
Name: Average Math Score, dtype: float64
----------------------
count    485490.000000
mean        472.006964
std          98.863310
min           6.445400
25%         405.044200
50%         475.477980
75%         542.831195
max         849.359740
Name: Average Reading Score, dtype: float64
----------------------
count    485490.000000
mean        475.808094
std          97.998470
min          25.158540
25%         405.762800
50%         475.512860
75%         546.381920
max         857.832900
Name: Average Science Score, dtype: float64
count    485490.000000
mean        629.039175
std         128.839689
min         105.496473
25%         536.312167
50%         627.318087
75%         721.099990
max        1123.882273
Name: Average Total Score, dtype: float64


In [15]:
pisa_clean.sample(10)

STIDSTD ST04Q01                ST13Q01 ST14Q01 ST14Q02 ST14Q03  \
93109     18640    Male      <ISCED level 3A>       No      No      No   
158207    12625    Male      <ISCED level 3A>      Yes     NaN     NaN   
130281     2508    Male      <ISCED level 3A>       No     Yes      No   
479280     3728    Male      <ISCED level 3A>       No      No      No   
283381     4681    Male      <ISCED level 3A>       No      No     NaN   
430034     1713    Male      <ISCED level 3A>      NaN     Yes     NaN   
25492      3342    Male      <ISCED level 3A>      NaN     NaN     Yes   
480202     4650    Male      <ISCED level 3A>      Yes     Yes      No   
418703     1159    Male      <ISCED level 3A>       No      No      No   
353322     3814  Female  <ISCED level 3B, 3C>       No      No     Yes   

       ST14Q04                ST17Q01 ST18Q01 ST18Q02  ... ST57Q03 ST57Q04  \
93109       No       <ISCED level 2>       No      No  ...     0.0     0.0   
158207     NaN      <ISCED level 3A>      NaN     Yes  ...     0.0     0.0   
130281      No      <ISCED level 3A>       No     Yes  ...     1.0     0.0   
479280      No      <ISCED level 3A>       No      No  ...     0.0     0.0   
283381     NaN  <ISCED level 3B, 3C>       No      No  ...     0.0     0.0   
430034     NaN      <ISCED level 3A>      NaN     Yes  ...     NaN     NaN   
25492      NaN      <ISCED level 3A>      NaN     NaN  ...     NaN     NaN   
480202      No      <ISCED level 3A>       No      No  ...     2.0     0.0   
418703      No       <ISCED level 1>       No      No  ...     NaN     NaN   
353322     NaN  <ISCED level 3B, 3C>       No      No  ...     0.0     0.0   

       ST57Q05  MMINS  LMINS  SMINS  Average Math Score  \
93109      0.0  350.0  350.0  350.0           580.47204   
158207     0.0  165.0  220.0  165.0           698.71476   
130281     0.0  180.0  135.0   90.0           635.38712   
479280     0.0  250.0  250.0  250.0           536.53996   
283381     0.0  250.0  200.0  100.0           543.78410   
430034     NaN    NaN    NaN    NaN           493.07526   
25492      NaN    NaN    NaN    NaN           403.88690   
480202     0.0  225.0  225.0  225.0           430.13710   
418703     NaN  270.0   90.0  450.0           386.98398   
353322     2.0  180.0  180.0   90.0           581.95202   

        Average Reading Score  Average Science Score  Average Total Score  
93109               602.08212              673.94628           812.324160  
158207              666.15750              710.49984           924.695620  
130281              594.86462              639.53748           835.058780  
479280              539.37000              587.78440           733.411440  
283381              552.76248              546.38194           728.904207  
430034              515.31164              537.61654           679.692900  
25492               435.35772              451.08166           564.737727  
480202              432.55090              460.59306           584.472720  
418703              311.29682              390.09698           491.787253  
353322              590.89122              559.43680           771.410687  

[10 rows x 25 columns]

### Make the variable names readable

In the original dataframe, the variable names are all codes and unreadable without the dictionary. Therefore, we will change them to accurately reflect the information within the column.

In [16]:
# Rename columns
pisa_clean.rename({'STIDSTD': 'Student ID',
                   'ST04Q01': 'Gender', 
                   'ST13Q01': 'Mother - Highest Schooling', 
                   'ST14Q01': 'Mother Qualifications - ISCED level 6',
                   'ST14Q02': 'Mother Qualifications - ISCED level 5A',
                   'ST14Q03': 'Mother Qualifications - ISCED level 5B',
                   'ST14Q04': 'Mother Qualifications - ISCED level 4',
                   'ST17Q01': 'Father - Highest Schooling',
                   'ST18Q01': 'Father Qualifications - ISCED level 6',
                   'ST18Q02': 'Father Qualifications - ISCED level 5A',
                   'ST18Q03': 'Father Qualifications - ISCED level 5B',
                   'ST18Q04': 'Father Qualifications - ISCED level 4',
                   'ST28Q01': 'How many books at home', 
                   'ST57Q01': 'Out-of-School Study Time - Homework',
                   'ST57Q02': 'Out-of-School Study Time - Guided Homework',
                   'ST57Q03': 'Out-of-School Study Time - Personal Tutor',
                   'ST57Q04': 'Out-of-School Study Time - Commercial Company',
                   'ST57Q05': 'Out-of-School Study Time - With Parent',
                   'MMINS': 'Learning time (minutes per week)- <Mathematics>',
                   'LMINS': 'Learning time (minutes per week)  - <test language>',
                   'SMINS': 'Learning time (minutes per week) - <Science>'}, axis='columns', inplace=True)

In [17]:
# Check to see if the columns are correctly renamed
list(pisa_clean)

['Student ID',
 'Gender',
 'Mother - Highest Schooling',
 'Mother Qualifications - ISCED level 6',
 'Mother Qualifications - ISCED level 5A',
 'Mother Qualifications - ISCED level 5B',
 'Mother Qualifications - ISCED level 4',
 'Father - Highest Schooling',
 'Father Qualifications - ISCED level 6',
 'Father Qualifications - ISCED level 5A',
 'Father Qualifications - ISCED level 5B',
 'Father Qualifications - ISCED level 4',
 'How many books at home',
 'Out-of-School Study Time - Homework',
 'Out-of-School Study Time - Guided Homework',
 'Out-of-School Study Time - Personal Tutor',
 'Out-of-School Study Time - Commercial Company',
 'Out-of-School Study Time - With Parent',
 'Learning time (minutes per week)- <Mathematics>',
 'Learning time (minutes per week)  - <test language>',
 'Learning time (minutes per week) - <Science>',
 'Average Math Score',
 'Average Reading Score',
 'Average Science Score',
 'Average Total Score']

### Education
Multiple columns with the single variable. 

http://uis.unesco.org/sites/default/files/documents/international-standard-classification-of-education-isced-2011-en.pdf

In [18]:
father_edu = pisa_clean.copy()

In [19]:
father_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'He did not complete <ISCED level 1>']] = pd.get_dummies(father_edu['Father - Highest Schooling'])

In [20]:
father_edu.rename({'He did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Father Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Father Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Father Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [21]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].astype(str)
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].astype(str)
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].astype(str)
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].astype(str)

In [22]:
father_edu['<ISCED level 0>'] = father_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
father_edu['<ISCED level 1>'] = father_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
father_edu['<ISCED level 2>'] = father_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
father_edu['<ISCED level 3>'] = father_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
father_edu['<ISCED level 4>'] = father_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
father_edu['<ISCED level 5>'] = father_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
father_edu['<ISCED level 6>'] = father_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [23]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
158964       13382    Male          <ISCED level 3A>    
402496        2280  Female          <ISCED level 3A>    
32466        10316    Male          <ISCED level 3A>    
254240       13651  Female          <ISCED level 3A>    
332630       16928  Female      <ISCED level 3B, 3C>    
59394         4129    Male          <ISCED level 3A>    
279791        1091  Female          <ISCED level 3A>    
445742        2148    Male          <ISCED level 3A>    
278800         100    Male          <ISCED level 3A>    
355500        1248  Female           <ISCED level 1>    

       Mother Qualifications - ISCED level 6  \
158964                                    No   
402496                                    No   
32466                                     No   
254240                                    No   
332630                                    No   
59394                                     No   
279791                                    No   
445742                                    No   
278800                                    No   
355500                                    No   

       Mother Qualifications - ISCED level 5A  \
158964                                     No   
402496                                     No   
32466                                      No   
254240                                    Yes   
332630                                    Yes   
59394                                      No   
279791                                     No   
445742                                    Yes   
278800                                    Yes   
355500                                     No   

       Mother Qualifications - ISCED level 5B  \
158964                                     No   
402496                                     No   
32466                                      No   
254240                                     No   
332630                                    Yes   
59394                                      No   
279791                                    NaN   
445742                                     No   
278800                                    NaN   
355500                                     No   

       Mother Qualifications - ISCED level 4 Father - Highest Schooling  \
158964                                    No          <ISCED level 3A>    
402496                                   Yes           <ISCED level 2>    
32466                                     No      <ISCED level 3B, 3C>    
254240                                    No          <ISCED level 3A>    
332630                                   NaN      <ISCED level 3B, 3C>    
59394                                     No           <ISCED level 1>    
279791                                   NaN          <ISCED level 3A>    
445742                                   NaN          <ISCED level 3A>    
278800                                   NaN          <ISCED level 3A>    
355500                                    No           <ISCED level 1>    

       <ISCED level 6>  <ISCED level 5>  ...  \
158964              No               No  ...   
402496              No               No  ...   
32466              NaN              NaN  ...   
254240              No  <ISCED level 5>  ...   
332630              No  <ISCED level 5>  ...   
59394               No               No  ...   
279791              No               No  ...   
445742              No               No  ...   
278800              No  <ISCED level 5>  ...   
355500              No               No  ...   

       Learning time (minutes per week) - <Science> Average Math Score  \
158964                                          NaN          458.72412   
402496                                          0.0          499.38462   
32466                                         280.0          451.09050   
254240                                        180.0          454.51784   
332630                                        

In [24]:
father_edu = pd.melt(father_edu, 
                     id_vars=['Student ID', 'Gender',
                            'Mother - Highest Schooling', 
                            'Mother Qualifications - ISCED level 6', 
                            'Mother Qualifications - ISCED level 5A',
                            'Mother Qualifications - ISCED level 5B',
                            'Mother Qualifications - ISCED level 4',
                            'How many books at home',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning time (minutes per week)- <Mathematics>',
                            'Learning time (minutes per week)  - <test language>',
                            'Learning time (minutes per week) - <Science>',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score'], var_name='Completed_edu - Father', value_name='Education - Father')


In [25]:
father_edu.sample(10)

Student ID  Gender Mother - Highest Schooling  \
1493628         527    Male          <ISCED level 3A>    
1273469        1686  Female          <ISCED level 3A>    
3580103        5999    Male      <ISCED level 3B, 3C>    
3243726       15084  Female           <ISCED level 1>    
4562516        3990  Female          <ISCED level 3A>    
1751516        4187  Female                        NaN   
361867         2418  Female           <ISCED level 2>    
297076          891    Male                        NaN   
495491         5259    Male          <ISCED level 3A>    
2827811         145    Male          <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
1493628                                   NaN   
1273469                                   NaN   
3580103                                    No   
3243726                                    No   
4562516                                    No   
1751516                                    No   
361867                                     No   
297076                                     No   
495491                                    NaN   
2827811                                    No   

        Mother Qualifications - ISCED level 5A  \
1493628                                    Yes   
1273469                                    NaN   
3580103                                     No   
3243726                                     No   
4562516                                    Yes   
1751516                                     No   
361867                                      No   
297076                                     Yes   
495491                                     NaN   
2827811                                     No   

        Mother Qualifications - ISCED level 5B  \
1493628                                    NaN   
1273469                                    NaN   
3580103                                    Yes   
3243726                                     No   
4562516                                    Yes   
1751516                                     No   
361867                                      No   
297076                                     NaN   
495491                                     Yes   
2827811                                    Yes   

        Mother Qualifications - ISCED level 4 How many books at home  \
1493628                                   NaN    More than 500 books   
1273469                                   Yes          26-100 books    
3580103                                    No          26-100 books    
3243726                                   NaN            0-10 books    
4562516                                   Yes    More than 500 books   
1751516                                   NaN            0-10 books    
361867                                    NaN           11-25 books    
297076                                     No           11-25 books    
495491                                    NaN          26-100 books    
2827811                                    No            0-10 books    

         Out-of-School Study Time - Homework  \
1493628                                  5.0   
1273469                                  1.0   
3580103                                  2.0   
3243726                                  2.0   
4562516                                  8.0   
1751516                                  1.0   
361867                                   NaN   
297076                                   7.0   
495491                                  10.0   
2827811                                  1.0   

         Out-of-School Study Time - Guided Homework  ...  \
1493628                                         2.0  ...   
1273469                                         0.0  ...   
3580103                                         0.0  ...   
3243726                                         1.0  ...   
4562516                                         0.0  ...   
1751516                                       

In [26]:
father_edu = father_edu[(father_edu['Education - Father'] != '0')] 
father_edu = father_edu[(father_edu['Education - Father'] != 'No')]

In [27]:
father_edu.sample(70)

Student ID  Gender             Mother - Highest Schooling  \
4322380         777  Female                       <ISCED level 2>    
2242987         224    Male                      <ISCED level 3A>    
221683          295    Male                      <ISCED level 3A>    
313850         3483    Male                                    NaN   
3665548       26529    Male                      <ISCED level 3A>    
2180583        3089  Female                                    NaN   
4832627        2235    Male  She did not complete <ISCED level 1>    
2284532       26870  Female                       <ISCED level 2>    
2381900        2257  Female                  <ISCED level 3B, 3C>    
1758186         913    Male                      <ISCED level 3A>    
387479         3951    Male                      <ISCED level 3A>    
3572020        2695  Female                      <ISCED level 3A>    
123210           39    Male  She did not complete <ISCED level 1>    
207645          745  Female                  <ISCED level 3B, 3C>    
3002644       15235    Male                      <ISCED level 3A>    
1869890         950  Female                      <ISCED level 3A>    
3621211        1393    Male                      <ISCED level 3A>    
88447         13978    Male                      <ISCED level 3A>    
1335811         922  Female                  <ISCED level 3B, 3C>    
3282916        1381    Male                       <ISCED level 1>    
276024         4362  Female                       <ISCED level 2>    
635672         4600    Male                       <ISCED level 2>    
327120        11418    Male                       <ISCED level 2>    
1250413         733  Female                                    NaN   
1069290        2297    Male                  <ISCED level 3B, 3C>    
3984270        4337  Female                  <ISCED level 3B, 3C>    
2233488         669    Male                      <ISCED level 3A>    
2114264        1409  Female                      <ISCED level 3A>    
55123          5140    Male                                    NaN   
1357753        3245    Male                      <ISCED level 3A>    
...             ...     ...                                    ...   
4281562        8392    Male                                    NaN   
3558688       14676    Male                      <ISCED level 3A>    
1915915        5069  Female                  <ISCED level 3B, 3C>    
282363         3663  Female                      <ISCED level 3A>    
1361123         893    Male                  <ISCED level 3B, 3C>    
1136339       19777  Female                      <ISCED level 3A>    
3417703        3031  Female                                    NaN   
452391         4061    Male                       <ISCED level 2>    
599749          161  Female                      <ISCED level 3A>    
366279         2370  Female                      <ISCED level 3A>    
634439         3367    Male                      <ISCED level 3A>    
3794549        6869    Male                      <ISCED level 3A>    
1243864        1222    Male                                    NaN   
2103452       15910  Female                                    NaN   
108720         1478  Female                       <ISCED level 2>    
1328347        3115    Male                       <ISCED level 2>    
3269935        2743    Male                       <ISCED level 2>    
2236652        3833  Female                                    NaN   
1833962        4606    Male                      <ISCED level 3A>    
425368         2593  Female                      <ISCED level 3A>    
3776845        5529  Female                      <ISCED level 3A>    
1250950        1270    Male                      <ISCED level 3A>    
2125877        8243  Female                      <ISCED level 3A>    
2154071         541    Male                      <ISCED level 3A>    
699635         2575  Female                       <ISCED level 2>    
1605833        3781

In [28]:
father_edu = father_edu[(father_edu['Completed_edu - Father'] != 'Father - Highest Schooling')] 

In [29]:
father_edu['Education - Father'].value_counts()

<ISCED level 3>    306459
<ISCED level 5>    185130
<ISCED level 2>     84329
<ISCED level 4>     77265
<ISCED level 1>     35938
<ISCED level 6>     29589
<ISCED level 0>     16535
Name: Education - Father, dtype: int64

In [30]:
father_edu.drop(columns=['Completed_edu - Father'], axis=1, inplace=True)

In [31]:
father_edu.sample(10)

Student ID  Gender             Mother - Highest Schooling  \
2259824        2162    Male  She did not complete <ISCED level 1>    
3505295       10852    Male                  <ISCED level 3B, 3C>    
2170948        1978    Male                      <ISCED level 3A>    
2811434         456    Male                       <ISCED level 2>    
3109657        7601  Female                      <ISCED level 3A>    
3772628        1312    Male                       <ISCED level 1>    
1379187        1053  Female                       <ISCED level 2>    
976434          712    Male                      <ISCED level 3A>    
1722677       25618    Male                  <ISCED level 3B, 3C>    
3815642        4742    Male                      <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
2259824                                    No   
3505295                                    No   
2170948                                   NaN   
2811434                                    No   
3109657                                    No   
3772628                                    No   
1379187                                   NaN   
976434                                     No   
1722677                                    No   
3815642                                    No   

        Mother Qualifications - ISCED level 5A  \
2259824                                     No   
3505295                                     No   
2170948                                    NaN   
2811434                                     No   
3109657                                     No   
3772628                                     No   
1379187                                    NaN   
976434                                     Yes   
1722677                                     No   
3815642                                    Yes   

        Mother Qualifications - ISCED level 5B  \
2259824                                     No   
3505295                                    Yes   
2170948                                    NaN   
2811434                                     No   
3109657                                     No   
3772628                                     No   
1379187                                    NaN   
976434                                      No   
1722677                                    Yes   
3815642                                     No   

        Mother Qualifications - ISCED level 4 How many books at home  \
2259824                                   NaN            0-10 books    
3505295                                    No         101-200 books    
2170948                                   NaN          26-100 books    
2811434                                    No    More than 500 books   
3109657                                   Yes            0-10 books    
3772628                                    No            0-10 books    
1379187                                   NaN          26-100 books    
976434                                     No          26-100 books    
1722677                                    No          26-100 books    
3815642                                    No          26-100 books    

         Out-of-School Study Time - Homework  \
2259824                                  3.0   
3505295                                  3.0   
2170948                                  5.0   
2811434                                  NaN   
3109657                                  6.0   
3772628                                 10.0   
1379187                                  1.0   
976434                                  28.0   
1722677                                  NaN   
3815642                                  NaN   

         Out-of-School Study Time - Guided Homework  ...  \
2259824                                         0.0  ...   
3505295                                         0.0  ...   
2170948                                         0.0  ...   
2811434                           

In [32]:
father_edu.duplicated().sum()

327461

In [33]:
# https://stackoverflow.com/questions/33042777/removing-duplicates-from-pandas-dataframe-with-condition-for-retaining-original
father_edu['Education - Father'] = father_edu['Education - Father'].astype('category')

father_edu['Education - Father'] = father_edu['Education - Father'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
father_edu.sort_values(['Education - Father'], inplace=True)
# father_edu_clean = father_edu.drop_duplicates('Education - Father')

In [35]:
father_edu.duplicated().sum()

327461

In [37]:
father_edu.shape

(1263067, 21)

### Mother Edu

In [47]:
mother_edu = father_edu.copy()

In [48]:
mother_edu.head()

Student ID  Gender Mother - Highest Schooling  \
4854888        4948  Female           <ISCED level 1>    
4854895        4955  Female           <ISCED level 2>    
4427204        2529    Male           <ISCED level 1>    
4490059        6551  Female           <ISCED level 1>    
4490057        6549  Female          <ISCED level 3A>    

        Mother Qualifications - ISCED level 6  \
4854888                                    No   
4854895                                    No   
4427204                                    No   
4490059                                    No   
4490057                                   NaN   

        Mother Qualifications - ISCED level 5A  \
4854888                                     No   
4854895                                     No   
4427204                                     No   
4490059                                     No   
4490057                                    Yes   

        Mother Qualifications - ISCED level 5B  \
4854888                                    NaN   
4854895                                    NaN   
4427204                                     No   
4490059                                     No   
4490057                                    NaN   

        Mother Qualifications - ISCED level 4 How many books at home  \
4854888                                    No         101-200 books    
4854895                                    No          26-100 books    
4427204                                    No            0-10 books    
4490059                                   NaN            0-10 books    
4490057                                   NaN            0-10 books    

         Out-of-School Study Time - Homework  \
4854888                                  2.0   
4854895                                  5.0   
4427204                                  2.0   
4490059                                  NaN   
4490057                                  4.0   

         Out-of-School Study Time - Guided Homework  ...  \
4854888                                         2.0  ...   
4854895                                         NaN  ...   
4427204                                         2.0  ...   
4490059                                         NaN  ...   
4490057                                         NaN  ...   

         Out-of-School Study Time - Commercial Company  \
4854888                                            0.0   
4854895                                            0.0   
4427204                                            3.0   
4490059                                            NaN   
4490057                                            3.0   

         Out-of-School Study Time - With Parent  \
4854888                                     0.0   
4854895                                     0.0   
4427204                                     1.0   
4490059                                     NaN   
4490057                                     4.0   

         Learning time (minutes per week)- <Mathematics>  \
4854888                                            315.0   
4854895                                              NaN   
4427204                                            135.0   
4490059                                              NaN   
4490057                                            360.0   

         Learning time (minutes per week)  - <test language>  \
4854888                                                NaN     
4854895                                                NaN     
4427204                                              180.0     
4490059                                                NaN     
4490057                                              360.0     

         Learning time (minutes per week) - <Science>  Average Math Score  \
4854888                                         450.0           441.04222   
4854895                                           NaN           486.22058   
4427204                                        

In [49]:
mother_edu['Mother - Highest Schooling'].value_counts()

<ISCED level 3A>                         661214
<ISCED level 3B, 3C>                     210837
<ISCED level 2>                          184488
<ISCED level 1>                           78980
She did not complete <ISCED level 1>      40916
Name: Mother - Highest Schooling, dtype: int64

In [50]:
pisa_clean['Mother - Highest Schooling'].value_counts()

<ISCED level 3A>                         236993
<ISCED level 3B, 3C>                      83048
<ISCED level 2>                           82614
<ISCED level 1>                           36556
She did not complete <ISCED level 1>      18768
Name: Mother - Highest Schooling, dtype: int64

In [51]:
mother_edu[['<ISCED level 1>', 
            '<ISCED level 2>', 
            '<ISCED level 3A>', 
            '<ISCED level 3B, 3C>', 
            'She did not complete <ISCED level 1>']] = pd.get_dummies(mother_edu['Mother - Highest Schooling'])

In [52]:
mother_edu.rename({'She did not complete <ISCED level 1>': '<ISCED level 0>',
                   '<ISCED level 3A>': '<ISCED level 3>', 
                   '<ISCED level 3B, 3C>': '<ISCED level 3>',
                   'Mother Qualifications - ISCED level 4': '<ISCED level 4>',
                   'Mother Qualifications - ISCED level 5A': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 5B': '<ISCED level 5>',
                   'Mother Qualifications - ISCED level 6': '<ISCED level 6>'}, axis='columns', inplace=True)

In [53]:
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].astype(str)
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].astype(str)
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].astype(str)
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].astype(str)

In [54]:
mother_edu['<ISCED level 0>'] = mother_edu['<ISCED level 0>'].replace(regex='1', value='<ISCED level 0>')
mother_edu['<ISCED level 1>'] = mother_edu['<ISCED level 1>'].replace(regex='1', value='<ISCED level 1>')
mother_edu['<ISCED level 2>'] = mother_edu['<ISCED level 2>'].replace(regex='1', value='<ISCED level 2>')
mother_edu['<ISCED level 3>'] = mother_edu['<ISCED level 3>'].replace(regex='1', value='<ISCED level 3>')
mother_edu['<ISCED level 4>'] = mother_edu['<ISCED level 4>'].replace(regex='Yes', value='<ISCED level 4>')
mother_edu['<ISCED level 5>'] = mother_edu['<ISCED level 5>'].replace(regex='Yes', value='<ISCED level 5>')
mother_edu['<ISCED level 6>'] = mother_edu['<ISCED level 6>'].replace(regex='Yes', value='<ISCED level 6>')

In [55]:
mother_edu.sample(10)

Student ID  Gender             Mother - Highest Schooling  \
3759451        1572    Male                      <ISCED level 3A>    
1354840         332  Female                       <ISCED level 1>    
1958613         411  Female                       <ISCED level 1>    
4061222        1628    Male                      <ISCED level 3A>    
1077517       10524    Male                      <ISCED level 3A>    
1197478        5110    Male  She did not complete <ISCED level 1>    
3707477        2986  Female                      <ISCED level 3A>    
3714981         849    Male                      <ISCED level 3A>    
3487970       15071  Female                      <ISCED level 3A>    
3793205        5525  Female                      <ISCED level 3A>    

         <ISCED level 6>  <ISCED level 5>  <ISCED level 5>  <ISCED level 4>  \
3759451               No  <ISCED level 5>               No               No   
1354840               No               No               No               No   
1958613               No               No               No              NaN   
4061222              NaN              NaN              NaN              NaN   
1077517  <ISCED level 6>              NaN              NaN              NaN   
1197478              NaN              NaN              NaN              NaN   
3707477               No               No  <ISCED level 5>               No   
3714981              NaN              NaN               No              NaN   
3487970               No               No  <ISCED level 5>  <ISCED level 4>   
3793205               No  <ISCED level 5>  <ISCED level 5>               No   

        How many books at home  Out-of-School Study Time - Homework  \
3759451    More than 500 books                                  4.0   
1354840           11-25 books                                   NaN   
1958613           11-25 books                                   NaN   
4061222          26-100 books                                   NaN   
1077517           11-25 books                                   1.0   
1197478           11-25 books                                   NaN   
3707477          26-100 books                                   NaN   
3714981           11-25 books                                   NaN   
3487970           11-25 books                                   2.0   
3793205           11-25 books                                   2.0   

         Out-of-School Study Time - Guided Homework  ...  Average Math Score  \
3759451                                         0.0  ...           643.79966   
1354840                                         NaN  ...           545.73142   
1958613                                         NaN  ...           507.40770   
4061222                                         NaN  ...           412.22154   
1077517                                         0.0  ...           374.28728   
1197478                                         NaN  ...           365.01790   
3707477                                         NaN  ...           533.58000   
3714981                                         NaN  ...           291.25252   
3487970                                         NaN  ...           448.98740   
3793205                                         NaN  ...           365.17370   

         Average Reading Score  Average Science Score  Average Total Score  \
3759451              573.21206              632.07758           830.962987   
1354840              545.29782              478.77658           705.179080   
1958613              506.61492              505.91198           675.780767   
4061222              397.02476              407.44124           542.969693   
1077517              348.34668              359.60462           485.508620   
1197478              277.69532              279.41068           429.047267   
3707477              525.44008              540.50728           711.035787   
3714981              265.10478              296.56846           381.392760   
348

In [56]:
mother_edu = pd.melt(mother_edu, 
                     id_vars=['Student ID', 'Gender',
                            'How many books at home',
                            'Out-of-School Study Time - Homework',
                            'Out-of-School Study Time - Guided Homework',
                            'Out-of-School Study Time - Personal Tutor',
                            'Out-of-School Study Time - Commercial Company',
                            'Out-of-School Study Time - With Parent',
                            'Learning time (minutes per week)- <Mathematics>',
                            'Learning time (minutes per week)  - <test language>',
                            'Learning time (minutes per week) - <Science>',
                            'Average Math Score',
                            'Average Reading Score',
                            'Average Science Score',
                            'Average Total Score',
                             'Education - Father'], var_name='Completed_edu - Mother', value_name='Education - Mother')


In [57]:
mother_edu.sample(10)

Student ID  Gender How many books at home  \
8355754         2336    Male          26-100 books    
9869631         6159    Male    More than 500 books   
6433971         1066    Male         201-500 books    
12580604        3858    Male           11-25 books    
2967268         5650    Male          26-100 books    
3900651         7807    Male          26-100 books    
1406379          147    Male           11-25 books    
2791937        11193    Male          26-100 books    
4210080         3202  Female            0-10 books    
4967581         3624  Female                    NaN   

          Out-of-School Study Time - Homework  \
8355754                                  18.0   
9869631                                   NaN   
6433971                                   NaN   
12580604                                  1.0   
2967268                                   1.0   
3900651                                   9.0   
1406379                                   NaN   
2791937                                   7.0   
4210080                                   5.0   
4967581                                   NaN   

          Out-of-School Study Time - Guided Homework  \
8355754                                          9.0   
9869631                                          0.0   
6433971                                          NaN   
12580604                                         1.0   
2967268                                          0.0   
3900651                                          1.0   
1406379                                          NaN   
2791937                                          2.0   
4210080                                          3.0   
4967581                                          NaN   

          Out-of-School Study Time - Personal Tutor  \
8355754                                         0.0   
9869631                                         0.0   
6433971                                         NaN   
12580604                                        1.0   
2967268                                         0.0   
3900651                                         0.0   
1406379                                         NaN   
2791937                                         0.0   
4210080                                         0.0   
4967581                                         NaN   

          Out-of-School Study Time - Commercial Company  \
8355754                                             0.0   
9869631                                             0.0   
6433971                                             NaN   
12580604                                            0.0   
2967268                                             3.0   
3900651                                             0.0   
1406379                                             NaN   
2791937                                             4.0   
4210080                                             0.0   
4967581                                             NaN   

          Out-of-School Study Time - With Parent  \
8355754                                      5.0   
9869631                                      0.0   
6433971                                      NaN   
12580604                                     2.0   
2967268                                      1.0   
3900651                                      0.0   
1406379                                      NaN   
2791937                                      1.0   
4210080                                      0.0   
4967581                                      NaN   

          Learning time (minutes per week)- <Mathematics>  \
8355754                                             250.0   
9869631                                             200.0   
6433971                                               NaN   
12580604                                             90.0   
2967268                                             200.0   
3900651                                             

In [59]:
mother_edu = mother_edu[(mother_edu['Education - Mother'] != '0')] 
mother_edu = mother_edu[(mother_edu['Education - Mother'] != 'No')]

In [60]:
mother_edu.sample(70)

Student ID  Gender How many books at home  \
1422014         7137    Male         101-200 books    
378156           760  Female         101-200 books    
5300291          813  Female         201-500 books    
5486146          972    Male           11-25 books    
290519          2334  Female          26-100 books    
9891867         4932    Male          26-100 books    
4151143         2495  Female           11-25 books    
3561942          913    Male         201-500 books    
9229092         4888  Female          26-100 books    
3686041          866    Male         101-200 books    
133015          4555    Male         101-200 books    
4414870        10700  Female          26-100 books    
3683392         1487    Male            0-10 books    
665319          4224  Female           11-25 books    
2032259          703    Male                    NaN   
5427636         6525  Female          26-100 books    
5646543         8202    Male           11-25 books    
850588         13897  Female         201-500 books    
578808          4837    Male                    NaN   
1248238          799    Male                    NaN   
3867695          381    Male                    NaN   
6061812         8638    Male           11-25 books    
566657          2598  Female         201-500 books    
8326523         5518  Female          26-100 books    
6205551         4733    Male            0-10 books    
6705996        33062    Male            0-10 books    
6232815        27231    Male           11-25 books    
9024436         3034  Female    More than 500 books   
8938177        28551    Male          26-100 books    
1167386         8472  Female            0-10 books    
...              ...     ...                    ...   
2772864         3408    Male         201-500 books    
935868         15661  Female            0-10 books    
4955170         1098  Female    More than 500 books   
570575          3994  Female         201-500 books    
2444123        17137    Male         201-500 books    
4662184        27169  Female           11-25 books    
2037354         1067    Male          26-100 books    
3824819         3815  Female           11-25 books    
3181056         2202  Female           11-25 books    
2803683         7312    Male           11-25 books    
12167016         634  Female                    NaN   
4395660          237  Female         101-200 books    
62193            697    Male          26-100 books    
5249348        28593    Male          26-100 books    
2332905         2166  Female            0-10 books    
6185493         4580  Female            0-10 books    
256630          2197    Male          26-100 books    
4828266         2736  Female         201-500 books    
1167078         9574  Female          26-100 books    
3741144         9210  Female                    NaN   
4968202         5751    Male          26-100 books    
224428           841  Female          26-100 books    
5364744        15936  Female         101-200 books    
637430         12757    Male          26-100 books    
2013896         1628    Male         101-200 books    
167557            68  Female         201-500 books    
5438977         1757    Male          26-100 books    
1012276        11862  Female         201-500 books    
1067040          582  Female    More than 500 books   
777341          2153  Female           11-25 books    

          Out-of-School Study Time - Homework  \
1422014                                   NaN   
378156                                    NaN   
5300291                                   NaN   
5486146                                   1.0   
290519                                   24.0   
9891867                                   NaN   
4151143                                   1.0   
3561942                                   NaN   
9229092                                   NaN   
3686041                                   4.0   
133015                                    5.0   
4414870    

In [61]:
mother_edu = mother_edu[(mother_edu['Completed_edu - Mother'] != 'Mother - Highest Schooling')] 

In [62]:
mother_edu['Education - Mother'].value_counts()

<ISCED level 3>    872051
<ISCED level 5>    596385
<ISCED level 4>    206129
<ISCED level 2>    184488
<ISCED level 6>     85227
<ISCED level 1>     78980
<ISCED level 0>     40916
Name: Education - Mother, dtype: int64

In [63]:
mother_edu.drop(columns=['Completed_edu - Mother'], axis=1, inplace=True)

In [64]:
mother_edu.sample(10)

Student ID  Gender How many books at home  \
2022508        9339  Female           11-25 books    
4329800        1097  Female            0-10 books    
9006170       21742  Female    More than 500 books   
4748353        2179  Female            0-10 books    
9350865         918  Female         201-500 books    
3320992        7982    Male            0-10 books    
8566382        1938    Male    More than 500 books   
9590490        5663    Male         201-500 books    
6327305        3074    Male            0-10 books    
5416083        1871  Female           11-25 books    

         Out-of-School Study Time - Homework  \
2022508                                  NaN   
4329800                                  NaN   
9006170                                 14.0   
4748353                                  6.0   
9350865                                 11.0   
3320992                                  1.0   
8566382                                  4.0   
9590490                                  NaN   
6327305                                  NaN   
5416083                                  NaN   

         Out-of-School Study Time - Guided Homework  \
2022508                                         NaN   
4329800                                         NaN   
9006170                                         0.0   
4748353                                         NaN   
9350865                                         0.0   
3320992                                         0.0   
8566382                                         4.0   
9590490                                         NaN   
6327305                                         NaN   
5416083                                         NaN   

         Out-of-School Study Time - Personal Tutor  \
2022508                                        NaN   
4329800                                        NaN   
9006170                                        0.0   
4748353                                        0.0   
9350865                                        0.0   
3320992                                        0.0   
8566382                                        4.0   
9590490                                        NaN   
6327305                                        NaN   
5416083                                        NaN   

         Out-of-School Study Time - Commercial Company  \
2022508                                            NaN   
4329800                                            NaN   
9006170                                            3.0   
4748353                                            0.0   
9350865                                            0.0   
3320992                                            0.0   
8566382                                            4.0   
9590490                                            NaN   
6327305                                            NaN   
5416083                                            NaN   

         Out-of-School Study Time - With Parent  \
2022508                                     NaN   
4329800                                     NaN   
9006170                                     4.0   
4748353                                     1.0   
9350865                                     3.0   
3320992                                     0.0   
8566382                                     4.0   
9590490                                     NaN   
6327305                                     NaN   
5416083                                     NaN   

         Learning time (minutes per week)- <Mathematics>  \
2022508                                              NaN   
4329800                                              NaN   
9006170                                            220.0   
4748353                                            250.0   
9350865                                            135.0   
3320992                                              0.0   
8566382                                            180.0   
959

In [65]:
mother_edu['Student ID'].duplicated()

1263071     False
1263074     False
1263083     False
1263085     False
1263090     False
1263091     False
1263093     False
1263095     False
1263111     False
1263112     False
1263121     False
1263141     False
1263144     False
1263146     False
1263161     False
1263166     False
1263168     False
1263172     False
1263182     False
1263187     False
1263191     False
1263196     False
1263201     False
1263203     False
1263208     False
1263221     False
1263225     False
1263228     False
1263229     False
1263237     False
            ...  
12628706     True
12628886     True
12629156     True
12629267     True
12629304     True
12629305     True
12629377     True
12629499     True
12629520     True
12629610     True
12629647     True
12629674     True
12629695     True
12629723     True
12629728     True
12629782     True
12629808     True
12629854     True
12629922     True
12629936     True
12629960     True
12629966     True
12630019     True
12630067     True
12630142  

In [66]:
mother_edu['Education - Mother'] = mother_edu['Education - Mother'].astype('category')

mother_edu['Education - Mother'] = mother_edu['Education - Mother'].cat.set_categories(['<ISCED level 0>', 
                                                                                        '<ISCED level 1>', 
                                                                                        '<ISCED level 2>', 
                                                                                        '<ISCED level 3>',
                                                                                        '<ISCED level 4>', 
                                                                                        '<ISCED level 5>', 
                                                                                        '<ISCED level 6>'], ordered=True)
mother_edu.sort_values(['Education - Mother'], inplace=True)
# mother_edu_clean = mother_edu.drop_duplicates('Education - Mother')

In [70]:
mother_edu['Student ID'].duplicated().sum()

3865334

In [71]:
mother_edu.shape

(3899140, 17)

In [69]:
mother_edu_clean.shape

(8, 17)

In [84]:
pisa_clean = mother_edu.copy()

In [85]:
pisa_clean.shape

(3899140, 17)

In [90]:
duplicates = pisa_clean['Student ID'].duplicated().index

Int64Index([12455800, 12475494, 12622678, 12533982, 12550604, 12504251,
            12443148, 12550877, 12603191, 12533984,
            ...
             2871314,  2871590,  6087111,  6149270,  6149251,  2871808,
             2871468,  6087113,  6087121,  2871728],
           dtype='int64', length=3899140)

In [114]:
for x in duplicates:
    if pisa_clean.iloc[x]['Education - Mother'] == '<ISCED level 0>':
        pisa_clean.drop(pisa_clean.index[[x]])

IndexError: single positional indexer is out-of-bounds

In [109]:
pisa_clean.iloc[:5]['Education - Mother']

12455800    <ISCED level 0>
12475494    <ISCED level 0>
12622678    <ISCED level 0>
12533982    <ISCED level 0>
12550604    <ISCED level 0>
Name: Education - Mother, dtype: category
Categories (7, object): [<ISCED level 0> < <ISCED level 1> < <ISCED level 2> < <ISCED level 3> < <ISCED level 4> < <ISCED level 5> < <ISCED level 6>]

In [119]:
pisa_clean.iloc[6149251]['Education - Mother']

IndexError: single positional indexer is out-of-bounds

In [ ]:
pisa_clean = mother_edu.copy()

In [ ]:
pisa_clean.shape

### 4. Missing values in relevant columns ('Mother Highest Schooling', 'Father Highest Schooling', 'How many books at home', 'Out-of-School Study Time - Homework', 'Out-of-School Study Time - Guided Homework', 'Out-of-School Study Time - Personal Tutor','Out-of-School Study Time - Commercial Company', 'Out-of-School Study Time - With Parent', 'Learning time (minutes per week)- Mathematics', 'Learning time (minutes per week)  - test language', 'Learning time (minutes per week) - Science')

Due to the amount of rows in this dataframe, we can comfortably remove the rows with missing values.

In [ ]:

pisa_clean.replace(["NaN"], np.nan, inplace = True)
pisa_clean.dropna(inplace = True)

In [ ]:
# Check to see if null values correctly dropped
pisa_clean

In [ ]:
# Confirm with smaller shape of dataframe
pisa_clean.shape

### 5. Create a total for Out-of-School Study Time and Learning time (minutes per week)



In [ ]:
# Create total of Out-of-School Study Time
pisa_clean['Out-of-School Study Time - Total'] = pisa_clean['Out-of-School Study Time - Homework'] + pisa_clean['Out-of-School Study Time - Guided Homework'] + pisa_clean['Out-of-School Study Time - Personal Tutor'] + pisa_clean['Out-of-School Study Time - Commercial Company'] + pisa_clean['Out-of-School Study Time - With Parent']

# Create total of Learning time (minutes per week)
pisa_clean['Learning time (minutes per week) - Total'] = pisa_clean['Learning time (minutes per week)- <Mathematics>'] + pisa_clean['Learning time (minutes per week)  - <test language>'] + pisa_clean['Learning time (minutes per week) - <Science>'] 
                                                                                                                 

In [ ]:
list(pisa_clean)

<a id='assessing2'></a>
### Final Assessment

In this section, we will do the preliminary visual and programmatic assessment of the pisa datasets to determine whether or not it holds any major quality or tidiness issues.

In [ ]:
pisa_clean

In [ ]:
pisa_clean.info()

<a id='sav'></a>
## Store

In [ ]:
# Store the clean DataFrame in a CSV file
pisa_clean.to_csv("pisa_df.csv", index=False, encoding='utf8')

<a id='conclusion'></a>
## Conclusion

<a id='references'></a>
## References